src/binaryContentRenderer.ts

```diff
@@ -0,0 +1,116 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All Rights Reserved.
+ */
+
+import * as nls from 'vscode-nls';
+const localize: nls.LocalizeFunc = nls.loadMessageBundle();
+
+import * as vscode from 'vscode';
+import * as sarif from 'sarif';
+
+/**
+ * Class used to render binary embedded content.
+ */
+export class BinaryContentRenderer {
+    private static bytesPerRow: number = 8;
+    private static headerRows: number = 4;
+    private static markdownTableDelimiterLength: number = '|'.length;
+    private static hexNumberPrefixLength: number = '0x'.length;
+    private static hexNumberLength: number = 'ff'.length;
+    private static contentLengthPerTableCell: number = BinaryContentRenderer.hexNumberPrefixLength + BinaryContentRenderer.hexNumberLength + BinaryContentRenderer.markdownTableDelimiterLength;
+    private static binaryDataMarkdownHeader: string = localize("embeddedContent.tableHeader", "|Offset|0|1|2|3|4|5|6|7|\r\n|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|");
+
+    /**
+     * Creates an instance of the binary content renderer.
+     * @param content The string contents to be rendered as markdown.
+     */
+    private constructor(private readonly content: string) {
+    }
+
+    /**
+     * Attempts to create an instance of the binary content renderer based on a SARIF log, run index and artifact index
+     * Returns undefined if the artifact contents cannot be found, or the content is not binary content.
+     * @param log The SARIF log.
+     * @param runIndex The run index.
+     * @param artifactIndex The artifact index.
+     */
+    public static tryCreateFromLog(log: sarif.Log, runIndex: number, artifactIndex: number): BinaryContentRenderer | undefined {
+        const run: sarif.Run | undefined = log.runs[runIndex];
+        if (!run) {
+            return undefined;
+        }
+
+        if (!run.artifacts) {
+            return undefined;
+        }
+
+        const artifact: sarif.Artifact = run.artifacts[artifactIndex];
+        if (!artifact) {
+            return undefined;
+        }
+
+        if (!artifact.location) {
+            return undefined;
+        }
+
+        if (!artifact.contents) {
+            return undefined;
+        }
+
+        if (!artifact.contents.binary) {
+            return undefined;
+        }
+
+        return new BinaryContentRenderer(artifact.contents.binary);
+    }
+
+    public renderAsMarkdown(displayFileName: string): string {
```

Comment:
It is odd that the file name is a parameter but the content is a member.  Why?

Reply:
I should really rename the parameter. This "renderer" knows how to take a "content string" and render it as markdown. The "file" it is representing is really just a "URI" that exists in the SARF log that may not even be reachable. This is really just a "display file name"

---
In reply to: [410408869](https://github.com/microsoft/sarif-vscode-extension/pull/252#discussion_r410408869) [](ancestors = 410408869)